Fix misaligned pointer access in command list

killerdevildog opened the pull request:
- Align command sizes to pointer boundaries in mu_push_command
- Prevents undefined behavior and crashes on ARM and with sanitizers
- Maintains full backwards compatibility with existing code
- Fixes issue #19

The fix ensures all commands in the command buffer are properly aligned to sizeof(void*) boundaries, preventing misaligned memory access that caused runtime errors with address sanitizers and crashes on some architectures.

This pull request addresses issue #19 by fixing misaligned pointer access in the microui command list. The changes align command sizes to pointer boundaries in mu_push_command, preventing undefined behavior, crashes on ARM architectures, and runtime errors with sanitizers while maintaining full backwards compatibility. To verify, the demo was tested on this branch (fix-issue-19-command-alignment), where LeakSanitizer detected memory leaks but no alignment issues, confirming resolution. Switching to master, rebuilding, and running the demo reproduced the original UndefinedBehaviorSanitizer errors (e.g., misaligned member access at lines 430, 431, 471, 472, and 515 in microui.c), validating the fix's necessity.

killerdevildog committed:
Fix misaligned pointer access in command list

- Align command sizes to pointer boundaries in mu_push_command
- Prevents undefined behavior and crashes on ARM and with sanitizers
- Maintains full backwards compatibility with existing code
- Fixes issue #19

The fix ensures all commands in the command buffer are properly aligned
to sizeof(void*) boundaries, preventing misaligned memory access that
caused runtime errors with address sanitizers and crashes on some
architectures.

teleprint-me commented:
Note that the issue is not completely resolved. ASAN won't flag it if the pointer is realigned and this ends up being more of a hot-fix instead of a proper patch. This hot-fix actually masks the root issue which can end up being problematic. You can see issue #80 for more details on why this is the case.

oreo639 commented:
Just to clarify, this PR adds padding (specifically necessary for the `mu_TextCommand`) as needed to avoid misalignment for the following `mu_Command`.

For context, the `sizeof(void *)` is done here because `alignof(mu_Command)` doesn't exist pre-C11 and `void *dst` is the value with the largest alignment in the union.
It is possible to implement `alignof()` using `offsetof(struct {char c; mu_Command a;}, a)` though, which is available in C89. (this is what Nuklear does)
https://github.com/oreo639/microui/commit/33d03909d3ebc384eb94ae4d1061fd5fff552b65

There is also https://github.com/rxi/microui/pull/54, which resolves the issue by splitting the text into a separate buffer (also avoiding the misalignment, at the cost of using more memory).

teleprint-me commented:
Yeah, I had to reference the [NIST C89 publication](https://nvlpubs.nist.gov/nistpubs/Legacy/FIPS/fipspub160.pdf) for *§ 3.3.2.3 Structure and Union Members*. It's the closest reference to ANSI C (besides K&R).

> With one exception, if a member of a union object is accessed after a value has been stored in a different member of the object, the behavior is implementation-defined. One special guarantee is made in order to simplify the use of unions: **If a union contains several structures that share a common initial sequence (see below), and if the union object currently contains one of these structures, it is permitted to inspect the common initial part of any of them**. Two 10 structures share a common initial sequence if corresponding members have compatible types (and, for bit-fields, the same widths) for a sequence of one or more initial members.

This means that each structure must have the same type and `mu_Command` still violates this rule and is still undefined behavior. Even after alignment occurs.

```c
typedef struct { int type, size; } mu_BaseCommand; // applied dry to reduce repetition.
typedef struct { mu_BaseCommand base; void *dst; } mu_JumpCommand; // aligned
typedef struct { mu_BaseCommand base; mu_Rect rect; } mu_ClipCommand; // ...
typedef struct { mu_BaseCommand base; mu_Rect rect; mu_Color color; } mu_RectCommand;
typedef struct { mu_BaseCommand base; mu_Font font; mu_Vec2 pos; mu_Color color; char str[1]; } mu_TextCommand;
typedef struct { mu_BaseCommand base; mu_Rect rect; int id; mu_Color color; } mu_IconCommand;

typedef union {
  int type; // misaligns structs
  mu_BaseCommand base; // violates alignment expected by compiler.
  mu_JumpCommand jump; // aligned
  mu_ClipCommand clip; // ...
  mu_RectCommand rect;
  mu_TextCommand text;
  mu_IconCommand icon;
} mu_Command;
```

In order for this to be a valid ANSI C compliant implementation, each structure must identify itself with its type as `int type`. This means the issue is still valid.

`mu_Command` violates the exception to the rule. The cited section from the specification provides an example for coinciding with the exception to the rule.

I understand that pointer realignment is a valid trick and is common in allocators and other implementations, but it is masking the undefined behavior instead of resolving the root cause of the issue. I've worked on code pertaining to these alignment tricks.

```c
uintptr_t memory_align_up(uintptr_t value, uintptr_t alignment) {
    assert(memory_is_power_of_two(alignment));

    if (value > UINTPTR_MAX - alignment + 1) {
        return UINTPTR_MAX; // Overflow guard
    }

    return (value + alignment - 1) & ~(alignment - 1);
}

uintptr_t memory_align_down(uintptr_t value, uintptr_t alignment) {
    assert(memory_is_power_of_two(alignment));
    return value & ~(alignment - 1);
}
```

Just for sake of completeness, the C89 specification does explicitly state this as UB on pg 201.

```
An object is assigned to an overlapping object (3.3.16.1).
``` 

`mu_BaseCommand` is the overlapping object within the union.

It's better to treat the cause rather than the symptom.

I'm not expecting anything to come of this. Just bringing attention to it.

awschult002 commented:
there has been a new community fork of this repository. the community is looking to add some new simple features. please feel free to try your PR over there. 

https://github.com/microui-community/microui

killerdevildog committed:
Fix C89 union compliance for mu_Command

Addressed feedback from PR review - the original fix was masking the
problem rather than fixing the root cause.

Changes:
- Removed bare 'int type' from mu_Command union. The union now only
  contains structs that share a common initial sequence (mu_BaseCommand),
  which is well-defined per C89 section 3.3.2.3.
- Kept alignment padding in mu_push_command() to handle variable-size
  text commands.
- Memory usage unchanged - still uses the same 256KB buffer.

Builds clean with ASan/UBSan - no more sanitizer warnings.

Fixes #19
Addresses #80

killerdevildog commented:
Hey, thanks for the detailed feedback @teleprint-me and @oreo639 - you were absolutely right that my original fix was just masking the problem.

I went back and dug into the C89 spec you referenced. The issue was that having a bare int type in the union violated the "common initial sequence" rule. So I removed it - the union now only has struct members that all start with mu_BaseCommand base.

The fix ended up being pretty straightforward:

Removed int type from the union (the C89 violation)
Kept the alignment padding in mu_push_command() (fixes the misalignment from variable-size text commands)
Everything still uses the same 256KB buffer - no extra memory needed
Only change for users is accessing type through cmd->base.type instead of cmd->type, which is well-defined since all the structs share that common first member.

Let me know if I missed anything!

killerdevildog commented:
**Heads up my commit has a breaking change:**

In render loops, `cmd->type` becomes `cmd->base.type`. Here's an example from my test:

```c
/* Access through common initial sequence - this should never cause UB */
int type = cmd->base.type;

switch (type) {
    case MU_COMMAND_CLIP:
        /* handle clip */
        break;
    case MU_COMMAND_RECT:
        /* handle rect */
        break;
    case MU_COMMAND_TEXT:
        /* handle text - access via cmd->text.str, etc. */
        break;
    case MU_COMMAND_ICON:
        /* handle icon */
        break;
}
```
